### PR TITLE
fix: mark the topolvm storageclass as default

### PR DESCRIPTION
**Pull request by nbalacha**
This commit marks the topolvm storageClass as the default
if there are no other storageClasses that are present or
is no other class is marked default.

Signed-off-by: N Balachandran <nibalach@redhat.com>

### CODE REVIEW

File: controllers/topolvm_storageclass.go [hunk 1]
```diff
@@ -85,10 +85,33 @@ func (s topolvmStorageClass) updateStatus(r *LVMClusterReconciler, ctx context.C
 	return nil
 }
 
-func getTopolvmStorageClasses(lvmCluster *lvmv1alpha1.LVMCluster) []*storagev1.StorageClass {
+func getTopolvmStorageClasses(r *LVMClusterReconciler, ctx context.Context, lvmCluster *lvmv1alpha1.LVMCluster) []*storagev1.StorageClass {
 	sc := []*storagev1.StorageClass{}
 	allowVolumeExpansion := true
 	volumeBindingMode := storagev1.VolumeBindingWaitForFirstConsumer
+	//	defaultStorageClass := false
```
Comment (sp98) [inline controllers/topolvm_storageclass.go:92]:
```suggestion
	// defaultStorageClass := false
```

File: controllers/topolvm_storageclass.go [hunk 2]
```diff
@@ -85,10 +85,33 @@ func (s topolvmStorageClass) updateStatus(r *LVMClusterReconciler, ctx context.C
...
+
+	// Mark the first lvmo storage class default if no other storageclasses exist on the cluster
+	scList := &storagev1.StorageClassList{}
+	err := r.Client.List(context.TODO(), scList)
+
+	if err != nil {
+		r.Log.Error(err, "failed to list storage classes. Not setting any storageclass as the default")
+	} else {
+		if len(scList.Items) == 0 {
+			//defaultStorageClass = true
```
Comment (sp98) [inline controllers/topolvm_storageclass.go:103]:
```suggestion
			// defaultStorageClass = true
```

**Review by sp98 — CHANGES_REQUESTED** [review]
nit

File: controllers/topolvm_storageclass.go [hunk 3]
```diff
@@ -106,6 +129,11 @@ func getTopolvmStorageClasses(lvmCluster *lvmv1alpha1.LVMCluster) []*storagev1.S
 				"csi.storage.k8s.io/fstype": TopolvmFilesystemType,
 			},
 		}
+		// reconcile will pick up any existing LVMO storage classes as well
+		if defaultStorageClassName == "" || defaultStorageClassName == storageClass.Name {
```
Comment (sp98) [inline controllers/topolvm_storageclass.go:133]: won't ` defaultStorageClassName == "" ` condition be enough even if the reconciler picks up existing lvmo storage classes ?

**Review by sp98 — CHANGES_REQUESTED** [review]

File: controllers/topolvm_storageclass.go (hunk 3 above)
Comment (nbalacha) [reply controllers/topolvm_storageclass.go:133]: This is to handle a scenario where multiple VGs may be supported.

**Review by Yuggupta27 — CHANGES_REQUESTED** [review]

File: controllers/topolvm_storageclass.go [hunk 4]
```diff
@@ -85,10 +85,33 @@ func (s topolvmStorageClass) updateStatus(r *LVMClusterReconciler, ctx context.C
...
 	allowVolumeExpansion := true
 	volumeBindingMode := storagev1.VolumeBindingWaitForFirstConsumer
+	//	defaultStorageClass := false
+	defaultStorageClassName := ""
+
+	// Mark the first lvmo storage class default if no other storageclasses exist on the cluster
+	scList := &storagev1.StorageClassList{}
+	err := r.Client.List(context.TODO(), scList)
+
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (nbalacha) [reply controllers/topolvm_storageclass.go:100]: updated to set a new var in the if block.

File: controllers/topolvm_storageclass.go [hunk 5]
```diff
@@ -85,15 +85,38 @@ func (s topolvmStorageClass) updateStatus(r *LVMClusterReconciler, ctx context.C
...
+	// Mark the first lvmo storage class default if no other storageclasses exist on the cluster
+	scList := &storagev1.StorageClassList{}
+	err := r.Client.List(ctx, scList)
+
+	if err != nil {
+		r.Log.Error(err, "failed to list storage classes. Not setting any storageclass as the default")
+		setDefaultStorageClass = false
+	} else {
+		if len(scList.Items) != 0 {
+			for _, sc := range scList.Items {
```
[the remark this answers is not in the document]
Comment (nbalacha) [reply controllers/topolvm_storageclass.go]: Done.

File: controllers/topolvm_storageclass.go [hunk 6]
```diff
@@ -85,15 +85,38 @@ func (s topolvmStorageClass) updateStatus(r *LVMClusterReconciler, ctx context.C
...
+				if v == "true" {
+					defaultStorageClassName = sc.Name
+					break
+				}
+			}
+		}
+	}
+	sc := []*storagev1.StorageClass{}
 	for _, deviceClass := range lvmCluster.Spec.Storage.DeviceClasses {
+		scName := "odf-lvm-" + deviceClass.Name
```
[the remark this answers is not in the document]
Comment (nbalacha) [reply controllers/topolvm_storageclass.go:114]: That would need to be done in a separate PR.

File: controllers/topolvm_storageclass.go [hunk 7]
```diff
@@ -85,10 +85,33 @@ func (s topolvmStorageClass) updateStatus(r *LVMClusterReconciler, ctx context.C
...
+func getTopolvmStorageClasses(r *LVMClusterReconciler, ctx context.Context, lvmCluster *lvmv1alpha1.LVMCluster) []*storagev1.StorageClass {
 	sc := []*storagev1.StorageClass{}
 	allowVolumeExpansion := true
 	volumeBindingMode := storagev1.VolumeBindingWaitForFirstConsumer
+	//	defaultStorageClass := false
+	defaultStorageClassName := ""
+
+	// Mark the first lvmo storage class default if no other storageclasses exist on the cluster
+	scList := &storagev1.StorageClassList{}
+	err := r.Client.List(context.TODO(), scList)
```
[the remark this answers is not in the document]
Comment (nbalacha) [reply controllers/topolvm_storageclass.go]: Done.

File: controllers/topolvm_storageclass.go [hunk 8]
```diff
@@ -85,10 +85,33 @@ func (s topolvmStorageClass) updateStatus(r *LVMClusterReconciler, ctx context.C
...
+
+	if err != nil {
+		r.Log.Error(err, "failed to list storage classes. Not setting any storageclass as the default")
+	} else {
+		if len(scList.Items) == 0 {
+			//defaultStorageClass = true
+		} else {
+			for _, sc := range scList.Items {
+				v := sc.Annotations["storageclass.kubernetes.io/is-default-class"]
```
[the remark this answers is not in the document]
Comment (nbalacha) [reply controllers/topolvm_storageclass.go]: Done.

File: controllers/topolvm_storageclass.go (hunk 2 above)
Comment (nbalacha) [reply controllers/topolvm_storageclass.go]: Removed.

File: controllers/topolvm_storageclass.go (hunk 1 above)
Comment (nbalacha) [reply controllers/topolvm_storageclass.go]: Removed.

**Review by sp98 — APPROVED** [review]

**Pull request merged**